Using zig cc to compile acton leads to illegal instruction

Using `zig cc` to compile `builtin/hash.c` leads to `SIGILL`.
```c
long $pointer_hash($WORD p) {
    long x;
    long y = (long)p;
    // bottom 3 or 4 bits are likely to be 0; rotate y by 4 to avoid
    //  excessive hash collisions for dicts and sets 
    y = (y >> 4) | (y << 60);
    x = (long)y;
    if (x == -1)
        x = -2;
    return x;
}
```

Specifically, it's the bit shifting line that is causing this which we can reasily reproduce with this test program that also outputs a binary view to ease understanding:
```c
#include <stdio.h>

// Assumes little endian
void printBits(size_t const size, void const * const ptr)
{
    unsigned char *b = (unsigned char*) ptr;
    unsigned char byte;
    int i, j;

    for (i = size-1; i >= 0; i--) {
        for (j = 7; j >= 0; j--) {
            byte = (b[i] >> j) & 1;
            printf("%u", byte);
        }
    }
    puts("");
}

int main(int argc, char **argv) {
        long y = 1;
        y = (y >> 4) | (y << 60);
        long a = y;

        printf("a: %lu\n", a);
        printBits(sizeof(a), &a);
        printf("\n");
        return 0;
}
```
I can compile and run this with gcc & zig cc like so:
```
kll@Boxy:~/terastream/acton/testzig$ gcc test.c -o outgcc && zig cc test.c -o outzig && ./outgcc && ./outzig
a: 1152921504606846976
0001000000000000000000000000000000000000000000000000000000000000

a: 1152921504606846976
0001000000000000000000000000000000000000000000000000000000000000

kll@Boxy:~/terastream/acton/testzig$ 
```

With a small initial value of y it works as per above but when y = 8, which is binary 1000 and we shift that 60 bits we fall off the side of our long int and the zig cc compiled one SIGILLs (is that a verb?):

```
kll@Boxy:~/terastream/acton/testzig$ gcc test.c -o outgcc && zig cc test.c -o outzig && ./outgcc && ./outzig
a: 9223372036854775808
1000000000000000000000000000000000000000000000000000000000000000

Illegal instruction
kll@Boxy:~/terastream/acton/testzig$ 
```

gdb
```c
#0  0x00000000002017c7 in main (argc=1, argv=0x7ffc4a93af68) at test.c:21
21		y = (y >> 4) | (y << 60);
(gdb) 
```

I don't know the C standard well enough to say what should happen here. Is `zig cc` just doing something bad here? Can we shift and "trim" off the excess in some way, as to avoid the SIGILL?

I asked in the zig repo: https://github.com/ziglang/zig/issues/13246

Response:
> zig cc by default sets -fsanitize-undefined, use -fno-sanitize=undefined to turn it off.

So we get similar errors with gcc if we turn that on, although it actually appears to have some fancy runtime error wrapping handler:
```
kll@Boxy:~/terastream/acton/testzig$ gcc -fsanitize=undefined z.c
z.c: In function ‘main’:
z.c:3:21: warning: left shift count >= width of type [-Wshift-count-overflow]
    3 |         long a = (1 << 100);
      |                     ^~
kll@Boxy:~/terastream/acton/testzig$ ./a.out 
z.c:3:14: runtime error: shift exponent 100 is too large for 32-bit type 'int'
```

Turning it off the zig program does not crash with SIGILL:
```
kll@Boxy:~/terastream/acton/testzig$ zig cc -fno-sanitize=undefined z.c
kll@Boxy:~/terastream/acton/testzig$ ./a.out 
```

Soooo, it seems zig is just damn careful by default, which is probably a good thing... this in turn means our code is sort of incorrect, right? I mean, we might be deliberately shifting things outside the width of our type but can we make that more explicit somehow?

--- a/builtin/hash.c
+++ b/builtin/hash.c
@@ -137,12 +137,12 @@ long $complex_hash($complex c) {
 }  
  
 long $pointer_hash($WORD p) {
-    long x;
-    long y = (long)p;
+    unsigned long x;
+    unsigned long y = (unsigned long)p;
     // bottom 3 or 4 bits are likely to be 0; rotate y by 4 to avoid
     //  excessive hash collisions for dicts and sets 
     y = (y >> 4) | (y << 60);
-    x = (long)y;
+    x = (unsigned long)y;
     if (x == -1)
         x = -2;
     return x;

